Invalid memory address or nil pointer dereference when running dep ensure

### What version of Go and `dep` are you using?

* `go go1.9rc2 linux/amd64`
* `dep v0.3.0-132-g2097d86`

### What `dep` command did you run?

`dep ensure`

### What did you expect to see?

No errors

### What did you see instead?

```
dep ensure
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x75a336]

goroutine 75 [running]:
github.com/golang/dep/internal/gps.stripVendor(0xc42019aff0, 0x46, 0x0, 0x0, 0xb06fa0, 0xc4203502d0, 0xc420619d58, 0xc420294001)
	/go/src/github.com/golang/dep/internal/gps/strip_vendor.go:12 +0x26
path/filepath.Walk(0xc42019aff0, 0x46, 0x8f5a48, 0x13, 0xc420619d78)
	/usr/local/go/src/path/filepath/path.go:401 +0x76
github.com/golang/dep/internal/gps.WriteDepTree.func1(0xc420168460, 0x18, 0xc42031abe0, 0xb0f480, 0xc42005b860, 0xc42024a120, 0x446501, 0xc4204b8160, 0xc4201ccb40, 0x2d, ...)
	/go/src/github.com/golang/dep/internal/gps/result.go:82 +0x44b
created by github.com/golang/dep/internal/gps.WriteDepTree
	/go/src/github.com/golang/dep/internal/gps/result.go:73 +0x242
```

My Gopkg.toml:

```
[[constraint]]                                                                                                                                                                                                                              
  name = "my-private-gitlab.com/group/project"
  branch = "master"
  source = "ssh://git@my-private-gitlab.com/group/project.git"

[[constraint]]
  name = "github.com/urfave/cli"

--- a/internal/gps/result.go
+++ b/internal/gps/result.go
@@ -79,7 +79,10 @@ func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool, logger *log
 			}
 
 			if sv {
-				filepath.Walk(to, stripVendor)
+				err := filepath.Walk(to, stripVendor)
+				if err != nil {
+					errCh <- errors.Wrapf(err, "failed to strip vendor from %s", p.Ident().ProjectRoot)
+				}
 			}
 			wg.Done()
 		}(p)

--- a/internal/gps/result.go
+++ b/internal/gps/result.go
@@ -71,11 +71,13 @@ func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool, logger *log
 	for _, p := range l.Projects() {
 		wg.Add(1)
 		go func(p LockedProject) {
+			defer wg.Done()
 			to := filepath.FromSlash(filepath.Join(basedir, string(p.Ident().ProjectRoot)))
 			logger.Printf("Writing out %s@%s", p.Ident().errString(), p.Version())
 
 			if err := sm.ExportProject(p.Ident(), p.Version(), to); err != nil {
 				errCh <- errors.Wrapf(err, "failed to export %s", p.Ident().ProjectRoot)
+				return
 			}
 
 			if sv {
@@ -84,7 +86,7 @@ func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool, logger *log
 					errCh <- errors.Wrapf(err, "failed to strip vendor from %s", p.Ident().ProjectRoot)
 				}
 			}
-			wg.Done()
+			return
 		}(p)
 	}
 

--- a/internal/gps/result.go
+++ b/internal/gps/result.go
@@ -86,7 +86,6 @@ func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool, logger *log
 					errCh <- errors.Wrapf(err, "failed to strip vendor from %s", p.Ident().ProjectRoot)
 				}
 			}
-			return
 		}(p)
 	}
 

--- a/internal/gps/strip_vendor.go
+++ b/internal/gps/strip_vendor.go
@@ -12,7 +12,7 @@ import (
 )
 
 func stripVendor(path string, info os.FileInfo, err error) error {
-	if err != nil && err != filepath.SkipDir {
+	if err != nil {
 		return err
 	}
 

--- a/internal/gps/strip_vendor.go
+++ b/internal/gps/strip_vendor.go
@@ -17,20 +17,28 @@ func stripVendor(path string, info os.FileInfo, err error) error {
 	}
 
 	if info.Name() == "vendor" {
-		if _, err := os.Lstat(path); err == nil {
-			if (info.Mode() & os.ModeSymlink) != 0 {
-				realInfo, err := os.Stat(path)
-				if err != nil {
-					return err
-				}
-				if realInfo.IsDir() {
-					return os.Remove(path)
-				}
+		if _, err := os.Lstat(path); err != nil {
+			return err
+		}
+
+		if (info.Mode() & os.ModeSymlink) != 0 {
+			realInfo, err := os.Stat(path)
+			if err != nil {
+				return err
 			}
-			if info.IsDir() {
-				return removeAll(path)
+			if realInfo.IsDir() {
+				return os.Remove(path)
 			}
 		}
+
+		if info.IsDir() {
+			if err := removeAll(path); err != nil {
+				return err
+			}
+			return filepath.SkipDir
+		}
+
+		return nil
 	}
 
 	return nil

--- a/internal/gps/strip_vendor.go
+++ b/internal/gps/strip_vendor.go
@@ -6,9 +6,16 @@
 
 package gps
 
-import "os"
+import (
+	"os"
+	"path/filepath"
+)
 
 func stripVendor(path string, info os.FileInfo, err error) error {
+	if err != nil && err != filepath.SkipDir {
+		return err
+	}
+
 	if info.Name() == "vendor" {
 		if _, err := os.Lstat(path); err == nil {
 			if (info.Mode() & os.ModeSymlink) != 0 {

--- a/internal/gps/strip_vendor_windows.go
+++ b/internal/gps/strip_vendor_windows.go
@@ -10,6 +10,10 @@ import (
 )
 
 func stripVendor(path string, info os.FileInfo, err error) error {
+	if err != nil && err != filepath.SkipDir {
+		return err
+	}
+
 	if info.Name() == "vendor" {
 		if _, err := os.Lstat(path); err == nil {
 			symlink := (info.Mode() & os.ModeSymlink) != 0

--- a/internal/gps/strip_vendor_windows.go
+++ b/internal/gps/strip_vendor_windows.go
@@ -40,7 +40,10 @@ func stripVendor(path string, info os.FileInfo, err error) error {
 				}
 
 			case dir:
-				return removeAll(path)
+				if err := removeAll(path); err != nil {
+					return err
+				}
+				return filepath.SkipDir
 			}
 		}
 	}

--- a/internal/gps/strip_vendor_windows.go
+++ b/internal/gps/strip_vendor_windows.go
@@ -10,7 +10,7 @@ import (
 )
 
 func stripVendor(path string, info os.FileInfo, err error) error {
-	if err != nil && err != filepath.SkipDir {
+	if err != nil {
 		return err
 	}
 

